In [8]:
import warnings
warnings.filterwarnings(action='ignore')
import os
import pandas as pd
pd.set_option('display.max_rows', 300)
pd.set_option('display.max_columns', 50)
pd.set_option('display.expand_frame_repr', True)
pd.set_option('display.width', 1000)
pd.set_option('max_colwidth', 800)
import numpy as np
from tqdm.notebook import tqdm

In [2]:
domair = pd.read_parquet('/data1/DATA/resource/mkt_data/tour_domair_train.parquet')
domhotel = pd.read_parquet('/data1/DATA/resource/mkt_data/tour_domhotel_train.parquet')

- 유입 날짜 레벨 데이터

In [3]:
print(domair.shape)
print(domhotel.shape)

(4560902, 32)
(7304602, 33)


In [4]:
print(domair.columns)

Index(['memberId', 'date', 'last_visit_avg_duration', 'last_visit_total_duration', 'last_visit_purchase_amount', 'last_visit_main_duration', 'last_visit_search_duration', 'last_visit_resv_duration', 'last_visit_confirm_duration', 'last_visit_pageviews', 'last_visit_uniq_prd', 'last_visit_device_type', 'last_visit_prdviews', 'last_visit_quality', 'last_visit_hour', 'last_visit_channel', 'last_visit_interval_days', 'total_visit_pageviews', 'total_visit_duration', 'total_visit_prdviews', 'total_purchase_count', 'total_purchase_amount', 'total_visit_count', 'total_visit_cycle', 'total_purchase_recency', 'User_gender', 'User_age', 'is_visit_next_7days', 'is_visit_next_14days', 'is_visit_next_21days', 'is_visit_next_cycle', 'User_relation_history'], dtype='object')


- 항공은 검색 후 바로 예약 페이지로 넘어가기 때문에 기존 퍼널처럼 detail란이 부재함.

In [5]:
print(domhotel.columns)

Index(['memberId', 'date', 'last_visit_avg_duration', 'last_visit_total_duration', 'last_visit_purchase_amount', 'last_visit_main_duration', 'last_visit_search_duration', 'last_visit_detail_duration', 'last_visit_resv_duration', 'last_visit_confirm_duration', 'last_visit_pageviews', 'last_visit_uniq_prd', 'last_visit_device_type', 'last_visit_prdviews', 'last_visit_quality', 'last_visit_hour', 'last_visit_channel', 'last_visit_interval_days', 'total_visit_pageviews', 'total_visit_duration', 'total_visit_prdviews', 'total_purchase_count', 'total_purchase_amount', 'total_visit_count', 'total_visit_cycle', 'total_purchase_recency', 'User_gender', 'User_age', 'is_visit_next_7days', 'is_visit_next_14days', 'is_visit_next_21days', 'is_visit_next_cycle', 'User_relation_history'], dtype='object')


In [9]:
display(domair.head())
display(domhotel.head())

memberId      date  last_visit_avg_duration  last_visit_total_duration  last_visit_purchase_amount  last_visit_main_duration  last_visit_search_duration  last_visit_resv_duration  last_visit_confirm_duration  last_visit_pageviews  last_visit_uniq_prd last_visit_device_type  last_visit_prdviews  last_visit_quality  last_visit_hour last_visit_channel  last_visit_interval_days  total_visit_pageviews  total_visit_duration  total_visit_prdviews  total_purchase_count  total_purchase_amount  total_visit_count  total_visit_cycle  total_purchase_recency User_gender  User_age  is_visit_next_7days  is_visit_next_14days  is_visit_next_21days  is_visit_next_cycle  User_relation_history
0  ++mJKtssuoagt365xW5taQ==  20220308                 0.000000                   0.000000                         0.0                  0.000000                    0.000000                  0.000000                     0.000000                     1                    0                      3                    0                   2                3                  4                         1                      8             14.683883                     2                     8              1141500.0                  3                1.0                       2           2         3                    1                     1                     1                    1                      5
1  +0+u8fV8jjDO1ktVU2KGJw==  20220420                 0.833839                   2.501517                         0.0                  2.450383                    0.051133                  0.000000                     0.000000                     2                    1                      1                    1                   4                2                  3                         1                     27             25.442617                   114                    27                    0.0                  6                1.4                     999           1         6                    1                     1                     1                    1                      4
2  +0+u8fV8jjDO1ktVU2KGJw==  20220502                 0.077150                   0.308600                         0.0                  0.000000                    0.308600                  0.000000                     0.000000                     2                    2                      1                    2                   3                3                  3                         1                     60             80.923150                   153                    60                    0.0                 11                1.9                     999           1         6                    1                     1                     1                    1                      4
3  +0+u8fV8jjDO1ktVU2KGJw==  20220711                 2.452637                 100.558117                   2955600.0                 36.036750                   16.853367                 24.435467                    23.232533                    66                   10                      1                   29                  89                2                  1                        19                    200            316.173500                   214                   200              3406400.0                 22                4.2                     999           1         6                    0                     0                     1                    1                      4
4  +2f4VDdf1ETLClvaxLLsfg==  20220607                 1.039580                   5.197900                         0.0                  0.799700                    1.187933                  0.000000                     3.210267                    10                    2                      3                    4                  38                2                  1                        49                     23             17.831183                    10                    23               427200.0       

memberId      date  last_visit_avg_duration  last_visit_total_duration  last_visit_purchase_amount  last_visit_main_duration  last_visit_search_duration  last_visit_detail_duration  last_visit_resv_duration  last_visit_confirm_duration  last_visit_pageviews  last_visit_uniq_prd last_visit_device_type  last_visit_prdviews  last_visit_quality  last_visit_hour last_visit_channel  last_visit_interval_days  total_visit_pageviews  total_visit_duration  total_visit_prdviews  total_purchase_count  total_purchase_amount  total_visit_count  total_visit_cycle  total_purchase_recency User_gender  User_age  is_visit_next_7days  is_visit_next_14days  is_visit_next_21days  is_visit_next_cycle  User_relation_history
0  +0lHzw4aMMpaR3PCL4zjFQ==  20220611                 1.067332                  10.673317                         0.0                  1.263083                    3.299967                    5.540867                  0.569400                          0.0                    16                    2                      1                    4                  43                3                  1                         0                     16             10.673317                     4                    16                    0.0                  1              999.0                     999           2         3                    0                     0                     0                    0                      4
1  +3mieO7U75y7DBAhp5kSCw==  20220407                 1.043875                   8.351000                         0.0                  0.000000                    0.000000                    7.848383                  0.502617                          0.0                    13                    1                      1                    5                  45                3                  1                        51                     26              9.792250                     7                    26                    0.0                  2               51.0                     999           2         3                    1                     1                     1                    1                      4
2  +3vHArO/qDKfA3qJRX9PRg==  20220119                 0.003650                   0.007300                         0.0                  0.000000                    0.000000                    0.007300                  0.000000                          0.0                     2                    1                      3                    1                   1                3                  2                        15                     17              4.898117                     4                    17                    0.0                  2               15.0                     999           2         3                    0                     0                     0                    0                      4
3  +4XFwHTNGAh10rXQzqV81Q==  20220227                 0.851842                   3.407367                         0.0                  0.000000                    0.000000                    3.407367                  0.000000                          0.0                     8                    1                      1                    2                  24                3                  3                         2                    333            285.499017                    85                   333               573798.0                 20                3.0                       5           0         0                    1                     1                     1                    1                      1
4  +4YPiNhQGOX3QgCYQ0LVjA==  20220222                 1.713606                  20.563267                         0.0                 16.399600                    1.426550                    2.737117                  0.000000                          0.0                    34                    4                      1                    4                  42             

In [18]:
domair = domair.replace(999, None)
domhotel = domhotel.replace(999, None)
domair.loc[domair['last_visit_interval_days'] == 0, 'last_visit_interval_days'] = None
domhotel.loc[domhotel['last_visit_interval_days'] == 0, 'last_visit_interval_days'] = None

In [6]:
print(domair.memberId.nunique())
print(domhotel.memberId.nunique())

1167710
2170201


In [14]:
print(domair.groupby('memberId')['date'].count().mean())
print(domhotel.groupby('memberId')['date'].count().mean())

3.905851624119002
3.365864267871962


평균 방문주기, 재방문 주기 등은 타 wiki에서 참고함

- 평균 방문 주기 함수


def get_date_avg_interval(buyDateList):
        
        """날짜 리스트 안의 날짜들을 정렬 후 각 날짜 간 interval 의 평균을 산출

        Args:
            buyDateList (list): 구매 날짜 혹은 방문 날짜들을 포함하는 list

        Returns:
            [Double]: 평균 방문/구매 주기 (days)
        """
        retBuyDateAvg = 0
        periodList = []

        if buyDateList:
            buyDateList = np.sort(np.array(buyDateList))
            buyDateListCnt = len(buyDateList)
            if buyDateListCnt == 1:
                return None
            else:
                for idx in range(0, buyDateListCnt - 1):
                    period = 0
                    stimestmp = datetime.strptime(buyDateList[idx], '%Y%m%d')
                    etimestmp = datetime.strptime(buyDateList[idx + 1], '%Y%m%d')
                    period = (etimestmp - stimestmp).days
                    periodList.append(period)

                retBuyDateAvg = np.round(np.mean(periodList), 1)

        return float(retBuyDateAvg)
        
- domair_visit_days = domair_df.groupby('memberId')['date'].apply(list).reset_index()
- domhotel_visit_days = domhotel_df.groupby('memberId')['date'].apply(list).reset_index()

- domair_visit_days['visit_interval'] = domair_visit_days['date'].apply(get_date_avg_interval)
- domhotel_visit_days['visit_interval'] = domhotel_visit_days['date'].apply(get_date_avg_interval)

- print('국내항공 평균 방문 주기 : ' ,domair_visit_days['visit_interval'].mean())
- print('국내숙박 평균 방문 주기 : ' ,domhotel_visit_days['visit_interval'].mean())

- print('국내항공 재방문 interval', domair_df['last_visit_interval_days'].mean())
- print('국내숙박 재방문 interval', domhotel_df['last_visit_interval_days'].mean())

- print('국내 항공 재방문 수치') 
- print(domair_df['is_visit_next_7days'].value_counts())
- print(domair_df[ 'is_visit_next_14days'].value_counts())
- print(domair_df['is_visit_next_21days'].value_counts())
- print(domair_df['is_visit_next_cycle'].value_counts())

- print('국내 항공 재방문 비중 %') 방문 단위 이므로, 재방문을 하는 유저 수 비중과는 다름
- print(domair_df['is_visit_next_7days'].value_counts()/domair_df.shape[0])
- print(domair_df[ 'is_visit_next_14days'].value_counts()/domair_df.shape[0])
- print(domair_df['is_visit_next_21days'].value_counts()/domair_df.shape[0])
- print(domair_df['is_visit_next_cycle'].value_counts()/domair_df.shape[0])

In [20]:
print(domair['is_visit_next_7days'].value_counts()/domair.shape[0] * 100)

1    60.484812
0    39.515188
Name: is_visit_next_7days, dtype: float64


### 국내항공
- 국내항공 전체 유입 유저: 1,167,710
- 국내항공 유저별 평균 방문일자: 3.9일
- 국내항공 평균 방문주기 :  17.68일
- 국내항공 재방문주기: 10.37일
- 국내항공 재방문 여부(방문레벨) => 유저레벨로 봐야할 듯.. 
    - 7일 내 재방문 O(1): 2,758,653(60.5%) / 7일 내 재방문 X(0): 1,802,249(39.5%)
    - 14일 내 재방문 O(1): 3,208,175(70.3%) / 14일 내 재방문 X(0): 1,352,727(29.7%)
    - 21 내 재방문 O(1): 3,430,990(75.2%) / 21일 내 재방문 X(0): 1,129,912(24.8%)
    - 평균 구매 주기(35일) 내 재방문 O(1): 3,662,051(80.3%) / 평균 구매 주기 내 재방문 X(0): 898,851(19.7%)
    
** 아무리 방문 레벨이라도 방문일자 기준 7일 내에 재방문하는 경우가 60% 이상? 약 8개월 간의 방문한 모든 유저 데이터의 방문 레벨의 로그에서 60% 이상의 경우가 재방문을 7일이내에 했다..?
- 자주 방문하는 일부 유저가 매우 자주 방문해서 로그를 쌓았고 그게 영향을 끼쳤다? => 그렇다고 소수가 해당 로그의 60%를 차지한다는 건 조금..
- 모수가 전체 유입 유저인데..? 따로 조건이 없는 걸까?
- 정말 사려고 하는 유저들만 방문한다? 그래서 고민은 하지만 웬만해서 오래 고민하지 않고 바로 구매한다?
- 어떤 걸로 유입되었는 지는 봐야겠지만 한정 이벤트라 방문하고 바로 구매할 수도 있을 듯.(기한이 있는 한정 프로모션 등)
 - 빠르게 구매를 결정하는 사람들의 유입채널이 어디인지?
- 일부 충성 고객들이 방문해서 꾸준히 구매를 하는 것인지 아니면 체리피커들이 해당 시즌에 프로모션으로 인해서 방문하여 간 것인지(체리피커라도 항공권같은 경우 1번 이상 방문하여 고민은 할 수 있다고 생각) 해당 부분은 user_relation_history 등을 이용해서 파악해보면 좋을 듯

### 국내숙박
- 국내숙박 전체 유입 유저: 2,170,201
- 국내숙박 유저별 평균 방문일자: 3.37일
- 국내숙박 평균 방문주기 :  23.18일
- 국내숙박 재방문주기 11.92일


=> 방문한 유저를 대상으로 추출을 진행한 것이기 때문에 비교적 방문, 구매 주기가 높을 수 밖에 없음. 또한 분포를 확인 해 보아야겠지만 앞단에 쏠려있는 경우로 예상됨. 모든 유저가 해당 서비스에서 매번 저렇게 짧은 리텐션 주기를 갖긴 어려울 것.